pru_firmware/src/main_pru0.c

```diff
@@ -0,0 +1,162 @@
+#define ENCODER_MEM_OFFSET  16
+#define CAM_MEM_OFFSET  (ENCODER_MEM_OFFSET + 1)
+#define PRU_SHAREDMEM ((volatile unsigned int *)(0x10000)) // shared memory with Cortex A8 (Linux)
+
+
+// IO definitions
+#define LED_BATT1 (1<<27)  // (red) on GPIO0 - not recommended since Linux overwrites this regularly
+#define LED_BATT2 (1<<11)  // (green) on GPIO0 - not recommended since Linux overwrites this regularly
+#define LED_BATT3 (1<<29)  // (green) on GPIO1 - not recommended since Linux overwrites this regularly
+#define LED_BATT4 (1<<26)  // (green) on GPIO0 - not recommended since Linux overwrites this regularly
+
+#define LED_RED (1<<2)  // on GPIO2
+#define LED_GRN (1<<3)  // on GPIO3
+
+#define LED_USR0 (1<<21)  // (blue) on GPIO1
+#define LED_USR1 (1<<22)  // (blue) on GPIO1
+#define LED_USR3 (1<<24)  // (blue) on GPIO1
+
+#define LED_BUSY_GPIO GPIO2
+#define LED_BUSY_IO LED_RED
+
+#define CAM_SI_GPIO GPIO3
+#define CAM_SI_IO (1 << 17)
+#define CAM_CLK_GPIO GPIO3
+#define CAM_CLK_IO (1 << 20)
+#define CAM_AO_INP = 0  // ADC INP (positive in) channel for camera in
+
+
+// ADC helper prototypes
+void adc_init();
+uint16_t adc_read();
+
+
+// The main (top-level) PRU code
+int main() {
+  // Pulse the LED on quickly to indicate startup.
+  // You may want to remove this code so the main camera read code start immediately.
+  LED_BUSY_GPIO.SETDATAOUT = LED_BUSY_IO;
+  adc_init();
+  CAM_SI_GPIO.OE &= (~CAM_SI_IO) & (~CAM_CLK_IO);  // set the SI and CLK pins as output (we assume LED directions set by Linux)
+  CAM_SI_GPIO.CLEARDATAOUT = CAM_SI_IO;  // "initialize" SI low
+  CAM_CLK_GPIO.CLEARDATAOUT = CAM_CLK_IO;  // "initialize" CLK low
+  PRU_SHAREDMEM[CAM_MEM_OFFSET] = 0;
+  __delay_cycles(100*1000*200);
+  LED_BUSY_GPIO.CLEARDATAOUT = LED_BUSY_IO;
+  
+  while (1) {
+    while (PRU_SHAREDMEM[CAM_MEM_OFFSET] == 0);  // loop until read command from host
+    LED_BUSY_GPIO.SETDATAOUT = LED_BUSY_IO;
+    
+    CAM_SI_GPIO.SETDATAOUT = CAM_SI_IO;  // set SI high
+    __delay_cycles(200);  // you definitely have margin to play with here, and for the other __delay_cycles
+    CAM_CLK_GPIO.SETDATAOUT = CAM_CLK_IO;  // rising edge of CLK with SI asserted
+    __delay_cycles(200);
+    CAM_SI_GPIO.CLEARDATAOUT = CAM_SI_IO;
+    __delay_cycles(200);
+    
+    uint8_t i;
+    for (i=0; i<128; i++) {
+      CAM_CLK_GPIO.CLEARDATAOUT = CAM_CLK_IO;
+      PRU_SHAREDMEM[CAM_MEM_OFFSET + 1 + i] = adc_read();  // read on the falling CLK edge
+      // no need for a __delay_cycles here, since adc_read() takes 7.2us
+      CAM_CLK_GPIO.SETDATAOUT = CAM_CLK_IO;
+      __delay_cycles(200);
+    }
+    CAM_CLK_GPIO.CLEARDATAOUT = CAM_CLK_IO;
+    
+    PRU_SHAREDMEM[CAM_MEM_OFFSET] = 0;  // reset to zero to indicate done
+    LED_BUSY_GPIO.CLEARDATAOUT = LED_BUSY_IO;
+  }
+}
+
+
+// Initializes the ADC in "fast" mode
+// Based on the old PRU code at https://github.com/ucb-ee192/SkeletonBeagle/blob/master/pru_firmware/src/main_pru0.c
+void adc_init() {
```

Comment:
Most likely people are going to play black magic here, but maybe this is ok, since if they really want the speed, they'll take the time to read the datasheet more carefully.

Reply:
Yeah, this was mainly code from previous years. I assume it does something critical. Added a few more comments.